protocol/verification.go

```diff
@@ -0,0 +1,118 @@
+package protocol
+
+import (
+	"bytes"
+
+	"github.com/coniks-sys/coniks-go/crypto/sign"
+	"github.com/coniks-sys/coniks-go/crypto/vrf"
+	"github.com/coniks-sys/coniks-go/merkletree"
+)
+
+func (res *ErrorResponse) Verify(string, []byte, sign.PublicKey) ErrorCode {
+	return res.Error
+}
+
+func verifyAuthPath(uname string,
+	ap *merkletree.AuthenticationPath,
+	str *merkletree.SignedTreeRoot) ErrorCode {
+
+	// verify VRF Index
+	vrfKey := vrf.PublicKey(str.Policies.VrfPublicKey)
+	if !vrfKey.Verify([]byte(uname), ap.LookupIndex, ap.VrfProof) {
+		return ErrorBadVRFProof
+	}
+
+	// verify commitment
+	if bytes.Equal(ap.LookupIndex, ap.Leaf.Index) && /* proof of inclusion */
+		!ap.Leaf.Commitment.Verify([]byte(uname), ap.Leaf.Value) {
+		return ErrorBadCommitment
+	}
+
+	// verify auth path
+	if !ap.Verify(str.TreeHash) {
+		return ErrorBadMapping
+	}
+
+	return Passed
+}
+
+func verifySTR(signKey sign.PublicKey,
+	str *merkletree.SignedTreeRoot,
+	savedSTR []byte) ErrorCode {
+
+	// verify STR's signature
+	if !signKey.Verify(str.Serialize(), str.Signature) {
+		return ErrorBadSignature
+	}
+
+	// verify hash chain
+	if str.Epoch > 0 && savedSTR != nil &&
+		!merkletree.VerifyHashChain(str.PreviousSTRHash, savedSTR) {
+		return ErrorBadSTR
+	}
+
+	return Passed
+}
+
+func (df *DirectoryProof) Verify(uname string, savedSTR []byte, signKey sign.PublicKey) ErrorCode {
```

Comment:
I'd prefer to keep the `DirectoryResponse` declaration closer to these methods, so maybe move it over here.